### PR TITLE
🚧Move and improve scene system to Velaptor

### LINKED ISSUE #506: 🚧Move and improve scene system to Velaptor
### Complete The Item Below

- [X] I have updated the title without removing the 🚧 emoji.

### Description

Take all of the code from the **VelaptorTesting** related to scenes and move it to the **Velaptor** project.

**Reason:**
This will give users the ability to help manage scenes in the game and help with making games easier.

This will also require changes to the testing application due to the code migrating to **Velaptor**.

### Acceptance Criteria

- [x] Scene code migrated to Velaptor
- [x] VelaptorTesting refactored for changes

### ToDo Items

- [X] Change type labels added to this issue.  Refer to the _**Change Type Labels**_ section below.
- [X] Priority label added to this issue.  Refer to the _**Priority Type Labels**_ section below.
- [X] Issue linked to the correct project _(if applicable)_.
- [X] Issue linked to the correct milestone _(if applicable)_.
- [x] Draft pull request created and linked to this issue _(only required with code changes)_.

### Issue Dependencies

_No response_

### Related Work

_No response_

### Additional Information:

**_<details closed><summary>Change Type Labels</summary>_**

| Change Type         | Label                |
|---------------------|----------------------|
| Bug Fixes           | `🐛bug` |
| Breaking Changes    | `🧨breaking changes` |
| New Feature         | `✨new feature` |
| Workflow Changes    | `workflow` |
| Code Doc Changes    | `🗒️documentation/code` |
| Product Doc Changes | `📝documentation/product` |
</details>

**_<details closed><summary>Priority Type Labels</summary>_**

| Priority Type       | Label             |
|---------------------|-------------------|
| Low Priority        | `low priority`    |
| Medium Priority     | `medium priority` |
| High Priority       | `high priority`   |
</details>

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct.

### PR DESCRIPTION
**Pull request by CalvinWilkinson**
<h1 style="font-weight:bold" align="center">Preview Feature Pull Request</h1>
<h2 style="font-weight:bold" align="center">📋Review Checklist📋</h2>

- [x] The **_head(source)_** branch for this pull request is a **_preview feature_** branch, with the correct naming syntax, in the following example:
  - A **_head(source)_** branch is the branch being merged into a target branch.
    <details closed><summary>Example</summary>

      ``` xml
      Syntax: preview/feature/<issue-num>-<description>
      Example: preview/feature/123-my-preview-feature
      ```
    </details>

- [x] The **_head(source)_** branch for this pull request is created from a **_preview release_** branch, with the correct naming syntax, in the following example:
  <details closed><summary>Example</summary>

    ``` xml
    Syntax: preview/v<major>.<minor>.<patch>-preview.<prev-num>
    Example: preview/v1.2.3-preview.4
    ```
  </details>

- [x] The **_base(target)_** branch for this pull request is a **_preview release_** branch, with the correct naming syntax, in the following example:
  - A **_base(target)_** branch is the branch that the **_head(source)_** branch is merging into.

    <details closed><summary>Example</summary>

      ``` xml
      Syntax: preview/v<major>.<minor>.<patch>-preview.<prev-num>
      Example: preview/v1.2.3-preview.4
      ```
    </details>

💡For more information on branching, refer to the project's [branching documentation](../../Documentation/Branching.md).

- [x] Associated issue exists and is linked to this pull request.
  - One issue per pull request.

- [x] Pull request title matches the title of the associated issue.

- [x] A `preview` label has been added to the pull request.

- [x] The labels attached to this PR match the labels attached to the associated issue.

- [x] Project assigned matches the project of the associated issue.

- [x] Milestone assigned matches the milestone of the associated issue.

- [x] My code follows the coding style of this project.
  - The style is enforced by the **_.editorconfig_** files in the project and displayed as warnings.  If there is an edge case, reach out, and let's discuss it.

- [x] I have manually tested my code changes to the best of my ability.

- [x] All tests passed locally.
  - This is required because unit tests might pass locally but not in the CI environment during the status check process or vice-versa.
  - Tests might pass on the developer's machine but not necessarily on the code reviewer's machine.
  - Status checks run for this pull request when merging into the **_preview/v\*.\*.\*-preview.\*_** branch.  These status checks run every time a change is pushed to the pull request.  These checks validate version syntax, tagging, builds, unit tests, and more.

### CODE REVIEW

**Comment by CalvinWilkinson** [comment]
> **Warning** Added a https://github.com/KinsonDigital/Velaptor/labels/%F0%9F%A7%A8breaking%20changes
> This was due to renaming the `IWindowActions` interface to `IWindowEvents`

**Comment by CalvinWilkinson** [comment]
> **Warning** Added a https://github.com/KinsonDigital/Velaptor/labels/%F0%9F%A7%A8breaking%20changes
> This was due to deleting the `IWindowProps` interface.  It was combined into the `IWindow` interface.

**Comment by CalvinWilkinson** [comment]
> **Note** Increased https://github.com/KinsonDigital/Velaptor/labels/%F0%9F%8F%8E%EF%B8%8Fperformance by converting the `FrameTime` struct to a `readonly record` struct.

File: Velaptor/Factories/IReactableFactory.cs [hunk 1]
```diff
@@ -35,6 +35,12 @@ internal interface IReactableFactory
...
     IPushReactable<BatchSizeData> CreateBatchSizeReactable();
 
+    /// <summary>
+    /// Creates a window size reactable.
+    /// </summary>
+    /// <returns>The reactable.</returns>
+    IPushReactable<WindowSizeData> CreateWindowSizeReactable();
+
     /// <summary>
     /// Creates a view port reactable.
```
Comment (kselena) [inline Velaptor/Factories/IReactableFactory.cs:45]: Spelling.  Refactor to make viewport one word.
```suggestion
    /// Creates a viewport reactable.
```

File: Velaptor/Scene/ISceneManager.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    /// Unloads the scene manager content and added scenes.
+    /// </summary>
+    public void UnloadContent();
+
+    /// <summary>
+    /// Returns a value indicating whether or not a scene exists that matches the given <paramref name="id"/>.
+    /// </summary>
+    /// <param name="id">The ID of the scene.</param>
+    /// <returns>
+    ///     <c>true</c> if a scened exists with the given id.
```
Comment (kselena) [inline Velaptor/Scene/ISceneManager.cs:80]: Spelling (scene and ID)
```suggestion
    ///     <c>true</c> if a scene exists with the given ID.
```

File: Velaptor/Scene/SceneManager.cs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    public void AddScene(IScene scene, bool setToActive = false)
+    {
+        if (SceneExists(scene.Id))
+        {
+            var midSection = string.IsNullOrEmpty(scene.Name)
+                ? $"with the ID '{scene.Id}'"
+                : $"'{scene.Name}' with the ID '{scene.Id}'";
+            var exceptionMsg = $"The scene {midSection} already exists.";
+
+            // TODO: Create custom exception for this
```
Comment (kselena) [inline Velaptor/Scene/SceneManager.cs:48]: Do you still need this TODO?
```suggestion
            // TODO: Create custom exception for this
```

File: Velaptor/UI/IWindow.cs [hunk 4]
```diff
@@ -5,15 +5,120 @@
...
+    /// </summary>
+    uint Height { get; set; }
+
+    /// <summary>
+    /// Gets or sets the value of how often the update and render calls are invoked in the value of hertz.
+    /// </summary>
+    int UpdateFrequency { get; set; }
+
+    /// <summary>
+    /// Gets or sets a value indicating whether or not the buffers should]
```
Comment (kselena) [inline Velaptor/UI/IWindow.cs:72]: Grammar (remove bracket)
```suggestion
    /// Gets or sets a value indicating whether or not the buffers should
```

**Review by kselena — CHANGES_REQUESTED** [review]

File: Velaptor/Scene/SceneManager.cs (hunk 3 above)
Comment (CalvinWilkinson) [reply Velaptor/Scene/SceneManager.cs:51]: New exception created and test updated

File: Velaptor/Scene/ISceneManager.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    /// <summary>
+    /// Adds the given scene and sets it as active or inactive.
+    /// </summary>
+    /// <param name="scene">The scene to add.</param>
+    /// <param name="setToActive">
+    ///     When set to <c>true</c>, the scene being added will be set to active and
+    ///     the all other scenes will bet set to inactive.
+    /// </param>
+    public void AddScene(IScene scene, bool setToActive);
+
```
Comment (KinsonDigitalAdmin) [inline Velaptor/Scene/ISceneManager.cs:44]: remove extra line

**Review by KinsonDigitalAdmin — CHANGES_REQUESTED** [review]

**Review by KinsonDigitalAdmin — DISMISSED** [review]

File: Velaptor/Scene/ISceneManager.cs [hunk 6]
```diff
@@ -29,11 +29,18 @@ public interface ISceneManager : IUpdatable, IDrawable, IDisposable
...
+    /// <remarks>The scene will not be activated when added using this method.</remarks>
+    public void AddScene(IScene scene);
+
+    /// <summary>
+    /// Adds the given scene and sets it as active or inactive.
+    /// </summary>
+    /// <param name="scene">The scene to add.</param>
     /// <param name="setToActive">
     ///     When set to <c>true</c>, the scene being added will be set to active and
     ///     the all other scenes will bet set to inactive.
```
Comment (kselena) [inline Velaptor/Scene/ISceneManager.cs:41]: Grammar
```suggestion
    ///     all other scenes will bet set to inactive.
```

**Review by kselena — CHANGES_REQUESTED** [review]

**Review by KinsonDigitalAdmin — APPROVED** [review]

**Review by kselena — APPROVED** [review]

**Pull request merged**